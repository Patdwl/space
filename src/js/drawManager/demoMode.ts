import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { isDebrisOff, isInViewOff, isPayloadOff, isRocketBodyOff } from '../colorManager/colorSchemeManager';

let demoModeLastTime = 0;
export const demoMode = () => {
  const { drawManager, mainCamera, objectManager, sensorManager, timeManager, orbitManager, satSet } = keepTrackApi.programs;

  if (objectManager?.isSensorManagerLoaded && sensorManager?.currentSensor[0]?.lat === null) return;
  if (timeManager.realTime - demoModeLastTime < settingsManager.demoModeInterval) return;

  drawManager.demoModeLast = timeManager.realTime;
  drawManager.demoModeSatellite = drawManager.demoModeSatellite === satSet.satData.length ? 0 : drawManager.demoModeSatellite;

  let satData = satSet.satData;
  for (let i = drawManager.demoModeSatellite; i < satData.length; i++) {
    const sat = satData[i];
    if (isPayloadOff(sat) || isRocketBodyOff(sat) || isDebrisOff(sat) || isInViewOff(sat) || sat.static || sat.missile) continue;

    const satScreenPositionArray = satSet.getScreenCoords(i, drawManager.pMatrix, mainCamera.camMatrix);
    if (
      satScreenPositionArray.error ||
      typeof satScreenPositionArray.x == 'undefined' ||
      typeof satScreenPositionArray.y == 'undefined' ||
      satScreenPositionArray.x > window.innerWidth ||
      satScreenPositionArray.y > window.innerHeight
    ) {
      continue;
    }

    drawManager.hoverBoxOnSat(i, satScreenPositionArray.x, satScreenPositionArray.y);
    orbitManager.setSelectOrbit(i);
    drawManager.demoModeSatellite = i + 1;
    return;
  }
};
