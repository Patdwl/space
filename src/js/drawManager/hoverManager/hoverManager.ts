import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { getEl } from '@app/js/lib/helpers';
import { SatObject } from '../../api/keepTrackTypes';
import { SpaceObjectType } from '../../api/SpaceObjectType';
import { spaceObjType2Str } from '../../lib/spaceObjType2Str';

let satHoverBoxNode1: HTMLDivElement;
let satHoverBoxNode2: HTMLDivElement;
let satHoverBoxNode3: HTMLDivElement;
let satHoverBoxDOM: HTMLDivElement;

let currentSearchSats;
let updateHoverDelay = 0;
let updateHoverDelayLimit = 3;
let updateHoverSatId: number;

export const init = () => {
  // NOTE: Reusing these cached value causes the hover menu to get stuck on or off
  // when the user clicks on a satellite. You need to getElementById every time.
  satHoverBoxNode1 = <HTMLDivElement>(<unknown>getEl('sat-hoverbox1'));
  satHoverBoxNode2 = <HTMLDivElement>(<unknown>getEl('sat-hoverbox2'));
  satHoverBoxNode3 = <HTMLDivElement>(<unknown>getEl('sat-hoverbox3'));
  satHoverBoxDOM = <HTMLDivElement>(<unknown>getEl('sat-hoverbox'));
};
export const hoverOverNothing = () => {
  const { drawManager } = keepTrackApi.programs;
  satHoverBoxDOM = <HTMLDivElement>(<unknown>getEl('sat-hoverbox'));
  if (satHoverBoxDOM.style.display === 'none' || !settingsManager.enableHoverOverlay) return false;
  const { objectManager, starManager } = keepTrackApi.programs;
  if (objectManager.isStarManagerLoaded) {
    if (starManager.isConstellationVisible === true && !starManager.isAllConstellationVisible) starManager.clearConstellations();
  }

  satHoverBoxDOM.style.display = 'none';
  drawManager.canvas.style.cursor = 'default';
  drawManager.isHoverBoxVisible = false;
  return true;
};
export const hoverOverSomething = (satId: number, satX: number, satY: number) => {
  const { drawManager, mainCamera, satSet } = keepTrackApi.programs;
  if (!mainCamera.isDragging && settingsManager.enableHoverOverlay) {
    // NOTE: The radar mesurement logic breaks if you call it a SatObject

    const sat = <any>satSet.getSat(satId);
    drawManager.isHoverBoxVisible = true;    

    init();    

    if (sat.static || sat.isRadarData) {
      staticObj(sat);
    } else if (sat.missile) {
      missile(sat);
    } else if (sat.sccNum > 0) {
      satObj(sat);
    } else {
      staticObj(sat);
    }

    const style = {
      display: 'block',
      left: `${satX + 20}px`,
      top: `${satY - 10}px`,
    };
    Object.assign(satHoverBoxDOM.style, style);

    drawManager.canvas.style.cursor = 'pointer';
  }
};
export const staticObj = (sat: any) => {
  if (sat.type === SpaceObjectType.LAUNCH_FACILITY) {
    launchFacility(sat);
  } else if (sat.isRadarData) {
    // TODO: This is a broken mess but only used offline
    radarData(sat);
  } else if (sat.type === SpaceObjectType.CONTROL_FACILITY) {
    controlFacility(sat);
  } else if (sat.type === SpaceObjectType.STAR) {
    star(sat);
  } else {
    const { objectManager, satellite, satSet } = keepTrackApi.programs;
    satHoverBoxNode1.textContent = sat.name;
    satHoverBoxNode2.innerHTML = spaceObjType2Str(sat.type) + satellite.distance(sat, satSet.getSat(objectManager.selectedSat)) + '';
    satHoverBoxNode3.textContent = '';
  }
};
export const missile = (sat: SatObject) => {
  satHoverBoxNode1.innerHTML = sat.name + '<br >' + sat.desc + '';
  satHoverBoxNode2.textContent = '';
  satHoverBoxNode3.textContent = '';
};
export const satObj = (sat: SatObject) => {
  if (!settingsManager.enableHoverOverlay) return;
  const { drawManager, objectManager, satellite, sensorManager, satSet } = keepTrackApi.programs;
  // Use this as a default if no UI
  if (settingsManager.disableUI) {
    satHoverBoxNode1.textContent = sat.name;
    satHoverBoxNode2.textContent = sat.sccNum;
    satHoverBoxNode3.textContent = objectManager.extractCountry(sat.country);
  } else {
    if (objectManager.isSensorManagerLoaded && sensorManager.currentSensor[0].lat != null && settingsManager.isShowNextPass && drawManager.isShowDistance) {
      satHoverBoxNode1.textContent = sat.name;
      satHoverBoxNode2.textContent = sat.sccNum;
      satHoverBoxNode3.innerHTML = satellite.nextpass(sat) + satellite.distance(sat, satSet.getSat(objectManager.selectedSat)) + '';
    } else if (drawManager.isShowDistance) {
      satHoverBoxNode1.textContent = sat.name;
      drawManager.sat2 = satSet.getSat(objectManager.selectedSat);
      if (typeof drawManager.sat2 !== 'undefined' && drawManager.sat2 !== null && sat !== drawManager.sat2) {
        const ric = satellite.sat2ric(sat, drawManager.sat2);
        satHoverBoxNode2.innerHTML = `${sat.sccNum}`;
        satHoverBoxNode3.innerHTML =
          `R: ${ric.position[0].toFixed(2)}km I: ${ric.position[1].toFixed(2)}km C: ${ric.position[2].toFixed(2)}km</br>` +
          `ΔR: ${ric.velocity[0].toFixed(2)}km/s ΔI: ${ric.velocity[1].toFixed(2)}km/s ΔC: ${ric.velocity[2].toFixed(2)}km/s</br>`;
      } else {
        satHoverBoxNode2.innerHTML = `${sat.sccNum}${satellite.distance(sat, drawManager.sat2)}`;
        if (settingsManager.isEciOnHover) {
        satHoverBoxNode3.innerHTML =
          'X: ' +
          sat.position.x.toFixed(2) +
          ' km' +
          ' Y: ' +
          sat.position.y.toFixed(2) +
          ' km' +
          ' Z: ' +
          sat.position.z.toFixed(2) +
          ' km' +
          '</br>' +
          'XDot: ' +
          sat.velocity.x.toFixed(2) +
          ' km/s' +
          ' YDot: ' +
          sat.velocity.y.toFixed(2) +
          ' km/s' +
          ' ZDot: ' +
          sat.velocity.z.toFixed(2) +
          ' km/s';
        } else {
          satHoverBoxNode3.innerHTML = '';
        }
      }
    } else if (objectManager.isSensorManagerLoaded && sensorManager.currentSensor[0].lat != null && settingsManager.isShowNextPass) {
      satHoverBoxNode1.textContent = sat.name;
      satHoverBoxNode2.textContent = sat.sccNum;
      satHoverBoxNode3.textContent = satellite.nextpass(sat);
    } else {
      satHoverBoxNode1.textContent = sat.name;
      satHoverBoxNode2.textContent = sat.sccNum;
      if (settingsManager.isEciOnHover) {
        satHoverBoxNode3.innerHTML =
        'X: ' +
        sat.position.x.toFixed(2) +
        ' Y: ' +
        sat.position.y.toFixed(2) +
        ' Z: ' +
        sat.position.z.toFixed(2) +
        '</br>X: ' +
        sat.velocity.x.toFixed(2) +
        ' Y: ' +
        sat.velocity.y.toFixed(2) +
        ' Z: ' +
        sat.velocity.z.toFixed(2);
      } else {
        satHoverBoxNode3.innerHTML = '';
      }
    }
  }
};
export const radarData = (sat: any) => {
  const { satellite, sensorManager, timeManager } = keepTrackApi.programs;

  satHoverBoxNode1.innerHTML = 'Measurement: ' + sat.mId + '</br>Track: ' + sat.trackId + '</br>Object: ' + sat.objectId;
  if (sat.missileComplex !== -1) {
    satHoverBoxNode1.insertAdjacentHTML('beforeend', '</br>Missile Complex: ' + sat.missileComplex);
    satHoverBoxNode1.insertAdjacentHTML('beforeend', '</br>Missile Object: ' + sat.missileObject);
  }
  if (parseInt(sat.sccNum) !== -1) satHoverBoxNode1.insertAdjacentHTML('beforeend', '</br>Satellite: ' + sat.sccNum);
  if (typeof sat.rae == 'undefined' && sensorManager.currentSensor !== sensorManager.defaultSensor) {
    sat.rae = satellite.eci2Rae(sat.t, sat.position, sensorManager.currentSensor[0]);
    sat.setRAE(sat.rae);
  }
  if (sensorManager.currentSensor !== sensorManager.defaultSensor) {
    let measurementDate = new Date(sat.t);
    satHoverBoxNode2.innerHTML =
      `JDAY: ${timeManager.getDayOfYear(measurementDate)} - ${measurementDate.toLocaleString('en-GB', { timeZone: 'UTC' }).slice(-8)}` +
      '</br>' +
      'R: ' +
      sat.rae.range.toFixed(2) +
      ' A: ' +
      sat.rae.az.toFixed(2) +
      ' E: ' +
      sat.rae.el.toFixed(2);
  } else {
    let measurementDate = new Date(sat.t);
    satHoverBoxNode2.innerHTML = `JDAY: ${timeManager.getDayOfYear(measurementDate)} - ${measurementDate.toLocaleString('en-GB', { timeZone: 'UTC' }).slice(-8)}`;
  }
  satHoverBoxNode3.innerHTML =
    'RCS: ' +
    parseFloat(sat.rcs).toFixed(2) +
    ' m^2 (' +
    (10 ** (parseFloat(sat.rcs) / 10)).toFixed(2) +
    ' dBsm)</br>Az Error: ' +
    sat.azError.toFixed(2) +
    '° El Error: ' +
    sat.elError.toFixed(2) +
    '°';
};
export const launchFacility = (sat: SatObject) => {
  const { objectManager, satellite, satSet } = keepTrackApi.programs;
  let launchSite = objectManager.extractLaunchSite(sat.name);
  satHoverBoxNode1.textContent = launchSite.site + ', ' + launchSite.sitec;
  satHoverBoxNode2.innerHTML = spaceObjType2Str(sat.type) + satellite.distance(sat, satSet.getSat(objectManager.selectedSat)) + '';
  satHoverBoxNode3.textContent = '';
};
export const controlFacility = (sat: SatObject) => {
  const { objectManager, satellite, satSet } = keepTrackApi.programs;
  satHoverBoxNode1.textContent = sat.name;
  satHoverBoxNode2.innerHTML = sat.country + satellite.distance(sat, satSet.getSat(objectManager.selectedSat)) + '';
  satHoverBoxNode3.textContent = '';
};
export const star = (sat: SatObject) => {
  const { starManager, objectManager } = keepTrackApi.programs;
  const constellationName = starManager.findStarsConstellation(sat.name);
  if (constellationName !== null) {
    satHoverBoxNode1.innerHTML = sat.name + '</br>' + constellationName;
  } else {
    satHoverBoxNode1.textContent = sat.name;
  }
  satHoverBoxNode2.innerHTML = 'Star';
  satHoverBoxNode3.innerHTML = 'RA: ' + sat.ra.toFixed(3) + ' deg </br> DEC: ' + sat.dec.toFixed(3) + ' deg';
  if (objectManager.lasthoveringSat !== sat.id && typeof sat !== 'undefined' && constellationName !== null) {
    starManager.drawConstellations(constellationName);
  }
};
export const planetariumView = (satId: number) => {
  const { drawManager, mainCamera } = keepTrackApi.programs;
  if (mainCamera.cameraType.current === mainCamera.cameraType.Planetarium && !settingsManager.isDemoModeOn) {
    satHoverBoxDOM.style.display = 'none';
    drawManager.canvas.style.cursor = satId === -1 ? 'default' : 'pointer';
    return true;
  }
  return false;
};
export const hoverBoxOnSat = (satId: number, satX: number, satY: number) => {  
  if (typeof satHoverBoxDOM === 'undefined' || satHoverBoxDOM === null) return;

  if (planetariumView(satId)) return;

  if (satId === -1) {
    hoverOverNothing();
  } else {
    hoverOverSomething(satId, satX, satY);
  }
};
// This is intentionally complex to reduce object creation and GC
// Splitting it into subfunctions would not be optimal
export const updateHover = () => { // NOSONAR
  const { drawManager, mainCamera, orbitManager, uiManager, searchBox, satSet, timeManager } = keepTrackApi.programs;  

  if (searchBox.isResultBoxOpen() && !settingsManager.disableUI && !settingsManager.lowPerf) {
    currentSearchSats = searchBox.getLastResultGroup();
    if (typeof currentSearchSats !== 'undefined') {
      currentSearchSats = currentSearchSats['sats'];
      if (drawManager.updateHoverI >= currentSearchSats.length) drawManager.updateHoverI = 0;
      for (let i = 0; drawManager.updateHoverI < currentSearchSats.length && i < 5; drawManager.updateHoverI++, i++) {
        orbitManager.updateOrbitBuffer(currentSearchSats[drawManager.updateHoverI].satId);
      }
    }
  }
  if (!settingsManager.disableUI && searchBox.isHovering()) {
    updateHoverSatId = uiManager.searchBox.getHoverSat();
    const satScreenPositionArray = satSet.getScreenCoords(updateHoverSatId, drawManager.pMatrix, mainCamera.camMatrix);
    try {
      hoverBoxOnSat(updateHoverSatId, satScreenPositionArray.x, satScreenPositionArray.y);
    } catch (e) {
      // Intentionally Empty
    }
  } else {
    // gl.readPixels in uiInput.getSatIdFromCoord creates a lot of jank
    // Earlier in the loop we decided how much to throttle updateHover
    // if we skip it this loop, we want to still draw the last thing
    // it was looking at

    if (1000 / timeManager.dt < 15) {
      updateHoverDelayLimit = settingsManager.updateHoverDelayLimitBig;
    } else if (1000 / timeManager.dt < 30) {
      updateHoverDelayLimit = settingsManager.updateHoverDelayLimitSmall;
    } else {
      if (updateHoverDelayLimit > 0) --updateHoverDelayLimit;
    }

    if (mainCamera.isDragging || settingsManager.isMobileModeEnabled) return;

    if (++updateHoverDelay >= updateHoverDelayLimit) {
      updateHoverDelay = 0;
      uiManager.uiInput.mouseSat = uiManager.uiInput.getSatIdFromCoord(mainCamera.mouseX, mainCamera.mouseY);
    }

    if (settingsManager.enableHoverOrbits) {
      if (uiManager.uiInput.mouseSat !== -1 && keepTrackApi.programs.satSet.satData[uiManager.uiInput.mouseSat].type !== SpaceObjectType.STAR) {
        orbitManager.setHoverOrbit(uiManager.uiInput.mouseSat);
      } else {
        orbitManager.clearHoverOrbit();
      }
      satSet.setHover(uiManager.uiInput.mouseSat);
    }
    if (settingsManager.enableHoverOverlay) {
      hoverBoxOnSat(uiManager.uiInput.mouseSat, mainCamera.mouseX, mainCamera.mouseY);
    }
  }
};

export const hoverManager = {
  init,
  updateHover,
};
