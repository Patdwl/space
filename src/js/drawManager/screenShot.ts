import { keepTrackApi } from '../api/keepTrackApi';

export const screenShot = () => {
  const { drawManager } = keepTrackApi.programs;
  if (!settingsManager.queuedScreenshot) {
    drawManager.resizeCanvas();
    settingsManager.queuedScreenshot = true;
  } else {
    let link = document.createElement('a');
    link.download = 'keeptrack.png';

    let d = new Date();
    let n = d.getUTCFullYear();
    let copyrightStr = !settingsManager.copyrightOveride ? `Â©${n} KEEPTRACK.SPACE` : '';

    link.href = watermarkedDataUrl(drawManager.canvas, copyrightStr);
    link.click();
    settingsManager.screenshotMode = false;
    settingsManager.queuedScreenshot = false;
    drawManager.resizeCanvas();
  }
};

export const watermarkedDataUrl = (canvas: HTMLCanvasElement, text: string) => {
  const tempCanvas = document.createElement('canvas');
  const tempCtx = tempCanvas.getContext('2d');
  let cw = (tempCanvas.width = canvas.width);
  let ch = (tempCanvas.height = canvas.height);
  tempCtx.drawImage(canvas, 0, 0);
  tempCtx.font = '24px nasalization';
  let textWidth = tempCtx.measureText(text).width;
  tempCtx.globalAlpha = 1.0;
  tempCtx.fillStyle = 'white';
  tempCtx.fillText(text, cw - textWidth - 30, ch - 30);

  if (settingsManager.classificationStr !== '') {
    tempCtx.font = '24px nasalization';
    textWidth = tempCtx.measureText(settingsManager.classificationStr).width;
    tempCtx.globalAlpha = 1.0;
    switch (settingsManager.classificationStr) {
      case 'Top Secret//SCI':
        tempCtx.fillStyle = '#fce93a';
        break;
      case 'Top Secret':
        tempCtx.fillStyle = '#ff8c00';
        break;
      case 'Secret':
        tempCtx.fillStyle = '#ff0000';
        break;
      case 'Confidential':
        tempCtx.fillStyle = '#0033a0';
        break;
      case 'CUI':
        tempCtx.fillStyle = '#512b85';
        break;
      case 'Unclassified':
        tempCtx.fillStyle = '#007a33';
        break;
      default:
        throw new Error('Invalid classification');
    }
    tempCtx.fillText(settingsManager.classificationStr, cw / 2 - textWidth, ch - 20);
    tempCtx.fillText(settingsManager.classificationStr, cw / 2 - textWidth, 34);
  }

  document.body.appendChild(tempCanvas);
  let image = tempCanvas.toDataURL();
  tempCanvas.parentNode.removeChild(tempCanvas);
  return image;
};
