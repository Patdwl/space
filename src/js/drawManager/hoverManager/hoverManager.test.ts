import { defaultSat, defaultSensor, keepTrackApiStubs } from '../../api/apiMocks';
import { keepTrackApi } from '../../api/keepTrackApi';
import { KeepTrackPrograms, SatObject, SettingsManager } from '../../api/keepTrackTypes';
import { SpaceObjectType } from '../../api/SpaceObjectType';
import { drawManager } from '../drawManager';
import * as hoverManager from './hoverManager';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });
declare const settingsManager: SettingsManager;

// @ponicode
describe('hoverManager.hoverBoxOnSat', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <canvas id="keeptrack-canvas"></canvas>
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
    <div id="sat-minibox"></div>
    `;
    drawManager.init();
  });
  test('0', () => {
    keepTrackApi.programs.mainCamera.cameraType.current = keepTrackApi.programs.mainCamera.cameraType.Planetarium;
    settingsManager.isDemoModeOn = false;
    let result: any = drawManager.hoverBoxOnSat(25544, 1000, 1000);
    expect(result).toMatchSnapshot();
    keepTrackApi.programs.mainCamera.cameraType.current = keepTrackApi.programs.mainCamera.cameraType.Default;
  });

  test('1', () => {
    let result: any = drawManager.hoverBoxOnSat(-1, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () => <SatObject>{ static: true };
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () =>
      <SatObject>(<unknown>{ isRadarData: true, setRAE: jest.fn(), rae: { range: 0, az: 0, el: 0 }, rcs: '0', azError: 0, elError: 0 });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () =>
      <SatObject>(<unknown>{ isRadarData: true, missileComplex: 1, setRAE: jest.fn(), rae: { range: 0, az: 0, el: 0 }, rcs: '0', azError: 0, elError: 0 });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () =>
      <SatObject>(<unknown>{ isRadarData: true, satId: 1, setRAE: jest.fn(), rae: { range: 0, az: 0, el: 0 }, rcs: '0', azError: 0, elError: 0 });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('6', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () => <SatObject>(<unknown>{
        static: true,
        type: SpaceObjectType.CONTROL_FACILITY,
        setRAE: jest.fn(),
        rae: { range: 0, az: 0, el: 0 },
        rcs: '0',
        azError: 0,
        elError: 0,
        position: { x: 0, y: 0, z: 0 },
        velocity: { total: 0, x: 0, y: 0, z: 0 },
      });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('7', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () => <SatObject>(<unknown>{
        static: true,
        type: SpaceObjectType.STAR,
        setRAE: jest.fn(),
        rae: { range: 0, az: 0, el: 0 },
        rcs: '0',
        azError: 0,
        elError: 0,
        position: { x: 0, y: 0, z: 0 },
        velocity: { total: 0, x: 0, y: 0, z: 0 },
        ra: 0,
        dec: 0,
      });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('8', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () => <SatObject>(<unknown>{
        missile: true,
        setRAE: jest.fn(),
        rae: { range: 0, az: 0, el: 0 },
        rcs: '0',
        azError: 0,
        elError: 0,
        position: { x: 0, y: 0, z: 0 },
        velocity: { total: 0, x: 0, y: 0, z: 0 },
      });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });

  test('8', () => {
    keepTrackApi.programs.satSet.getSatExtraOnly = () => <SatObject>(<unknown>{
        satId: 1,
        setRAE: jest.fn(),
        rae: { range: 0, az: 0, el: 0 },
        rcs: '0',
        azError: 0,
        elError: 0,
        position: { x: 0, y: 0, z: 0 },
        velocity: { total: 0, x: 0, y: 0, z: 0 },
      });
    let result: any = drawManager.hoverBoxOnSat(5, 1000, 1000);
    expect(result).toMatchSnapshot();
  });
});
// @ponicode
describe('hoverManager.updateHover', () => {
  test('0', () => {
    let result: any = hoverManager.updateHover();
    expect(result).toMatchSnapshot();
  });
});

describe('hoverManager.hoverOverNothing', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
  `;
    hoverManager.init();
  });
  it('should handle hovering over empty space', () => {
    keepTrackApi.programs.drawManager.isHoverBoxVisible = true;
    keepTrackApi.programs.objectManager.isStarManagerLoaded = true;
    settingsManager.enableHoverOverlay = true;
    let result: boolean = hoverManager.hoverOverNothing();
    expect(result).toBe(true);
  });
  it('should fail if the hover box is not visible', () => {
    keepTrackApi.programs.drawManager.isHoverBoxVisible = false;
    settingsManager.enableHoverOverlay = true;
    let result: boolean = hoverManager.hoverOverNothing();
    expect(result).toBe(false);
  });
  it('should fail if the hover overlay is not enabled', () => {
    keepTrackApi.programs.drawManager.isHoverBoxVisible = true;
    settingsManager.enableHoverOverlay = false;
    let result: boolean = hoverManager.hoverOverNothing();
    expect(result).toBe(false);
  });
});

describe('hoverManager.hoverOverSomething', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
  `;
    hoverManager.init();
  });
  it('should handle normal satellites', () => {
    settingsManager.enableHoverOverlay = true;
    hoverManager.hoverOverSomething(0, 0, 0);

    const satHoverBoxDOM = document.getElementById('sat-hoverbox');
    expect(satHoverBoxDOM).toMatchSnapshot();
  });
  it('should handle statics', () => {
    settingsManager.enableHoverOverlay = true;
    const tempSatSet = keepTrackApi.programs.satSet.getSat;
    keepTrackApi.programs.satSet.getSat = () => <SatObject>(<unknown>{ ...defaultSat, static: true });
    hoverManager.hoverOverSomething(10, 0, 0);
    const satHoverBoxDOM = document.getElementById('sat-hoverbox');
    expect(satHoverBoxDOM).toMatchSnapshot();
    keepTrackApi.programs.satSet.getSat = tempSatSet;
  });
  it('should handle missiles', () => {
    settingsManager.enableHoverOverlay = true;
    const tempSatSet = keepTrackApi.programs.satSet.getSat;
    keepTrackApi.programs.satSet.getSat = () => <SatObject>(<unknown>{ ...defaultSat, missile: true });
    hoverManager.hoverOverSomething(10, 0, 0);

    const satHoverBoxDOM = document.getElementById('sat-hoverbox');
    expect(satHoverBoxDOM).toMatchSnapshot();
    keepTrackApi.programs.satSet.getSat = tempSatSet;
  });
});

describe('hoverManager.staticObj', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
  `;
    hoverManager.init();
  });
  it('should handle launch facilities', () => {
    hoverManager.staticObj({ ...defaultSat, static: true, type: SpaceObjectType.LAUNCH_FACILITY });
  });
  it('should handle radar data', () => {
    hoverManager.staticObj({ ...defaultSat, static: true, isRadarData: true, setRAE: jest.fn(), rae: { range: 0, az: 0, el: 0 }, azError: 0, elError: 0 });
  });
  it('should handle control facilities', () => {
    hoverManager.staticObj({ ...defaultSat, static: true, type: SpaceObjectType.CONTROL_FACILITY });
  });
  it('should handle stars', () => {
    hoverManager.staticObj({ ...defaultSat, static: true, type: SpaceObjectType.STAR, ra: 0, dec: 0 });
  });
  it('should handle other statics', () => {
    hoverManager.staticObj({ ...defaultSat, static: true, type: SpaceObjectType.LAUNCH_AGENCY });
  });
});

describe('hoverManager.satObj', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
  `;
    hoverManager.init();
  });
  it('should handle normal satellites', () => {
    hoverManager.satObj({ ...defaultSat, static: false });
  });
  it('should handle ui being disabled', () => {
    settingsManager.disableUI = true;
    hoverManager.satObj({ ...defaultSat, static: false });
    settingsManager.disableUI = false;
  });
  it('should handle a sensor being loaded', () => {
    keepTrackApi.programs.objectManager.isSensorManagerLoaded = true;
    keepTrackApi.programs.sensorManager.currentSensor = [defaultSensor];
    settingsManager.isShowNextPass = true;
    keepTrackApi.programs.drawManager.isShowDistance = true;
    hoverManager.satObj({ ...defaultSat, static: false });
  });
  it('should handle showing distance with a second satellite', () => {
    keepTrackApi.programs.objectManager.isSensorManagerLoaded = false;
    keepTrackApi.programs.drawManager.isShowDistance = true;
    const tempGetSat = keepTrackApi.programs.satSet.getSat;
    keepTrackApi.programs.satSet.getSat = () => defaultSat;
    hoverManager.satObj({ ...defaultSat, static: false });
    keepTrackApi.programs.satSet.getSat = tempGetSat;
  });
  it('should handle showing distance without a second satellite', () => {
    keepTrackApi.programs.objectManager.isSensorManagerLoaded = false;
    keepTrackApi.programs.drawManager.isShowDistance = true;
    const tempGetSat = keepTrackApi.programs.satSet.getSat;
    keepTrackApi.programs.satSet.getSat = () => null;
    hoverManager.satObj({ ...defaultSat, static: false });
    keepTrackApi.programs.satSet.getSat = tempGetSat;
  });
  it('should handle showing the next pass', () => {
    keepTrackApi.programs.objectManager.isSensorManagerLoaded = true;
    keepTrackApi.programs.drawManager.isShowDistance = false;
    keepTrackApi.programs.sensorManager.currentSensor = [defaultSensor];
    hoverManager.satObj({ ...defaultSat, static: false });
  });
});

describe('hoverManager.updateHover', () => {
  beforeAll(() => {
    document.body.innerHTML = `
    <div id="sat-hoverbox"></div>
    <div id="sat-hoverbox1"></div>
    <div id="sat-hoverbox2"></div>
    <div id="sat-hoverbox3"></div>
  `;
    hoverManager.init();
  });
  it('should do something if the result box is open', () => {
    settingsManager.disableUI = false;
    settingsManager.lowPerf = false;
    keepTrackApi.programs.searchBox.isResultBoxOpen = () => true;
    keepTrackApi.programs.mainCamera.isDragging = false;
    settingsManager.isMobileModeEnabled = false;
    keepTrackApi.programs.satSet.satData[0] = { ...defaultSat, type: SpaceObjectType.STAR };
    hoverManager.updateHover();
    keepTrackApi.programs.satSet.satData[0] = { ...defaultSat, type: SpaceObjectType.LAUNCH_AGENCY };
    hoverManager.updateHover();
  });
  it('should show the search box', () => {
    settingsManager.disableUI = false;
    keepTrackApi.programs.searchBox.isHovering = () => true;
    const tempUi = keepTrackApi.programs.uiManager;
    keepTrackApi.programs.uiManager = <any>{
      searchBox: {
        getHoverSat: () => defaultSat.id,
      },
    };
    hoverManager.updateHover();
    keepTrackApi.programs.uiManager = tempUi;
  });
});
