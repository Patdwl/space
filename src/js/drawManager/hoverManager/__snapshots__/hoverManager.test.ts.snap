// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoverManager.hoverBoxOnSat 0 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 1 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 2 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 3 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 4 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 5 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 6 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 7 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 8 1`] = `undefined`;

exports[`hoverManager.hoverBoxOnSat 8 2`] = `undefined`;

exports[`hoverManager.hoverOverSomething should handle missiles 1`] = `
<div
  id="sat-hoverbox"
  style="display: block; left: 20px; top: -10px;"
/>
`;

exports[`hoverManager.hoverOverSomething should handle normal satellites 1`] = `
<div
  id="sat-hoverbox"
  style="display: block; left: 20px; top: -10px;"
/>
`;

exports[`hoverManager.hoverOverSomething should handle statics 1`] = `
<div
  id="sat-hoverbox"
  style="display: block; left: 20px; top: -10px;"
/>
`;

exports[`hoverManager.updateHover 0 1`] = `undefined`;
